File: csharp/Resources/OidcClientsResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Environment OIDC client Wave 3 operations
    /// </summary>
    public class OidcClientsResource
    {
        private readonly AuthdogClient _client;

        public OidcClientsResource(AuthdogClient client)
        {
            _client = client;
        }

        private static string Path(string tenantId, string applicationId, string environmentId) =>
            $"/v1/tenants/{tenantId}/applications/{applicationId}/environments/{environmentId}/oidc-clients";

        public Task<JObject> ListAsync(string tenantId, string applicationId, string environmentId) =>
            _client.RequestAsync<JObject>(HttpMethod.Get, Path(tenantId, applicationId, environmentId));

        public JObject List(string tenantId, string applicationId, string environmentId) =>
            ListAsync(tenantId, applicationId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> RegisterAsync(string tenantId, string applicationId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, Path(tenantId, applicationId, environmentId), body);

        public JObject Register(string tenantId, string applicationId, string environmentId, object body) =>
            RegisterAsync(tenantId, applicationId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> UpdateAsync(
            string tenantId,
            string applicationId,
            string environmentId,
            string clientId,
            object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Patch,
                $"{Path(tenantId, applicationId, environmentId)}/{clientId}",
                body);

        public JObject Update(
            string tenantId,
            string applicationId,
            string environmentId,
            string clientId,
            object body) =>
            UpdateAsync(tenantId, applicationId, environmentId, clientId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId, string applicationId, string environmentId, string clientId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{Path(tenantId, applicationId, environmentId)}/{clientId}");

        public JObject Delete(string tenantId, string applicationId, string environmentId, string clientId) =>
            DeleteAsync(tenantId, applicationId, environmentId, clientId).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/OidcClientsResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Environment OIDC client operations.
 */
public final class OidcClientsResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create an OIDC clients helper.
     * @param clientParam Authdog client
     */
    public OidcClientsResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List OIDC clients for a project environment.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @param environmentId environment ID
     * @return clients envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode list(final String tenantId,
                         final String applicationId,
                         final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                path(tenantId, applicationId, environmentId),
                null, null, JsonNode.class);
    }

    /**
     * Register an OIDC client.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @param environmentId environment ID
     * @param body request body
     * @return created client envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode register(final String tenantId,
                             final String applicationId,
                             final String environmentId,
                             final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                path(tenantId, applicationId, environmentId),
                body, null, JsonNode.class);
    }

    /**
     * Update an OIDC client.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @param environmentId environment ID
     * @param clientId client ID
     * @param body request body
     * @return updated client envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode update(final String tenantId,
                           final String applicationId,
                           final String environmentId,
                           final String clientId,
                           final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PATCH",
                path(tenantId, applicationId, environmentId)
                        + "/" + clientId,
                body, null, JsonNode.class);
    }

    /**
     * Delete an OIDC client.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @param environmentId environment ID
     * @param clientId client ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId,
                           final String applicationId,
                           final String environmentId,
                           final String clientId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                path(tenantId, applicationId, environmentId)
                        + "/" + clientId,
                null, null, JsonNode.class);
    }

    /**
     * Build the OIDC clients path prefix.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @param environmentId environment ID
     * @return path prefix
     */
    private String path(final String tenantId,
                        final String applicationId,
                        final String environmentId) {
        return "/v1/tenants/" + tenantId + "/applications/"
                + applicationId + "/environments/" + environmentId
                + "/oidc-clients";
    }
}
